recipe/session/accessToken.go

```diff
@@ -33,80 +35,199 @@ type accessTokenInfoStruct struct {
 	timeCreated             uint64
 }
 
-func getInfoFromAccessToken(jwtInfo ParsedJWTInfo, jwtSigningPublicKey string, doAntiCsrfCheck bool) (*accessTokenInfoStruct, error) {
+func getInfoFromAccessToken(jwtInfo ParsedJWTInfo, jwks keyfunc.JWKS, doAntiCsrfCheck bool) (*accessTokenInfoStruct, error) {
+	var payload map[string]interface{}
 
-	err := verifyJWT(jwtInfo, jwtSigningPublicKey)
-	if err != nil {
-		return nil, errors.TryRefreshTokenError{
-			Msg: err.Error(),
+	if jwtInfo.Version == 3 {
+		parsedToken, parseError := jwt.Parse(jwtInfo.RawTokenString, jwks.Keyfunc)
+		if parseError != nil {
+			return nil, sterrors.TryRefreshTokenError{
+				Msg: parseError.Error(),
+			}
+		}
+
+		if parsedToken.Valid {
+			claims, ok := parsedToken.Claims.(jwt.MapClaims)
+			if !ok {
+				return nil, sterrors.TryRefreshTokenError{
+					Msg: "Invalid JWT claims",
+				}
+			}
+
+			// Convert the claims to a key-value pair
+			claimsMap := make(map[string]interface{})
+			for key, value := range claims {
+				claimsMap[key] = value
+			}
+
+			payload = claimsMap
+		}
+	} else {
+		keys := []interface{}{}
+
+		for _, value := range jwks.ReadOnlyKeys() {
+			keys = append(keys, value)
+		}
+
+		for _, key := range keys {
+			/**
+			For each key we need to create a jwks structure to use with the verification library
+			{
+				keys: [...]
+			}
+			*/
+			keysTemp := [1]interface{}{key}
+			jwksTemp := map[string]interface{}{
+				"keys": keysTemp,
+			}
+
+			jsonString, marshalError := json.Marshal(jwksTemp)
+			if marshalError != nil {
+				return nil, sterrors.TryRefreshTokenError{
+					Msg: "Invalid JWK response",
+				}
+			}
+
+			jwksToUse, jwksError := keyfunc.NewJSON(jsonString)
```

Comment:
You're right, it does fail if the KID is missing in the created jwks. Updated the check to see if the error is due to signature or the kid missing (in either case it will continue with the loop instead of returning the error)

Reply:
Right I see what you mean, ive refactored the code to use the keys directly as opposed to relying on the JWK parsing. This way it would not throw an error if the kid isnt found in the JWT or throw if the kid isnt found in the JWKS I was forming.